### PR TITLE
Pareto set: Hypervolume

### LINKED ISSUE #181: Pareto set: hypervolume
As a user, I want functionality available to calculate the hypervolume of a Pareto front given the cell bounds, so that I can easily find the hypervolume when defining multi-objective acquisition functionality

### PR DESCRIPTION
**Pull request by satrialoka**
### Feature Implemented

This is a feature for Pareto functionality needed for Multi-objective BO, the code was ported from [GPFlowopt's pareto class](https://github.com/GPflow/GPflowOpt/blob/master/gpflowopt/pareto.py). This PR is small PR derived from previously large PR #126 and resolve #181
 
### Approach

Inside pareto.py `Pareto` class, this function is implemented:

1. `hypervolume_indicator` function: calculates hypervolume indicator given reference point 

## Status

Need review

## Notes

### CODE REVIEW

File: trieste/utils/pareto.py [hunk 1]
```diff
@@ -105,3 +105,32 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
         upper = tf.concat([range_ + 1, pf_ext_idx[::-1, 1:][: pf_ext_idx[-1, 0]]], axis=-1)
 
         return BoundedVolumes(lower, upper)
+
+    def hypervolume_indicator(self, reference: tf.Tensor) -> tf.Tensor:
+        """
+        Method to calculate the hypervolume indicator
+        The hypervolume indicator is the volume of the dominated region.
```
Comment (joelberkeley) [inline trieste/utils/pareto.py]: You don't need "method to .."

File: tests/unit/test_pareto.py [hunk 2]
```diff
@@ -111,3 +111,22 @@ def test_pareto_2d_bounds() -> None:
...
+            [0.9572, 0.6557],
+            [0.4854, 0.0357],
+            [0.8003, 0.8491],
+            [0.1419, 0.9340],
+        ]
+    )
+
+    pareto_2d = Pareto(objectives)
+
+    npt.assert_allclose(pareto_2d.hypervolume_indicator(tf.constant([2.0, 2.0])), 3.3878, rtol=1e-2)
```
Comment (joelberkeley) [inline tests/unit/test_pareto.py]: how did you calculate this?

File: trieste/utils/pareto.py [hunk 3]
```diff
@@ -105,3 +105,32 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
...
+
+        :param reference: a reference point to use, with shape [D].
+            Should be equal or bigger than the anti-ideal point of the Pareto set
+            for comparing results across runs, the same reference point must be used.
```
Comment (joelberkeley) [inline trieste/utils/pareto.py]: Is this meant to be
```
            Should be equal or bigger than the anti-ideal point of the Pareto set.
            For comparing results across runs, the same reference point must be used.
```
or
```
            Should be equal or bigger than the anti-ideal point of the Pareto set
            for comparing results across runs. The same reference point must be used.
```

File: trieste/utils/pareto.py [hunk 4]
```diff
@@ -105,3 +105,32 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
...
+        :return: hypervolume indicator (the higher the better)
```
Comment (joelberkeley) [inline trieste/utils/pareto.py]: `(the higher the better)` adds context that might not be relevant for a given user. Do you think it's necessary?

File: trieste/utils/pareto.py [hunk 5]
```diff
@@ -105,3 +105,32 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
         upper = tf.concat([range_ + 1, pf_ext_idx[::-1, 1:][: pf_ext_idx[-1, 0]]], axis=-1)
 
         return BoundedVolumes(lower, upper)
+
+    def hypervolume_indicator(self, reference: tf.Tensor) -> tf.Tensor:
```
Comment (joelberkeley) [inline trieste/utils/pareto.py]:
```suggestion
    def hypervolume_indicator(self, reference: TensorType) -> TensorType:
```

File: trieste/utils/pareto.py [hunk 6]
```diff
@@ -105,3 +105,32 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
...
+        """
+        Method to calculate the hypervolume indicator
+        The hypervolume indicator is the volume of the dominated region.
+
+        :param reference: a reference point to use, with shape [D].
```
Comment (joelberkeley) [inline trieste/utils/pareto.py:114]: could you add an assert_shapes that checks the shape of this - particularly that D matches the `front` or `bounds`

File: trieste/utils/pareto.py (hunk 3 above)
Comment (joelberkeley) [inline trieste/utils/pareto.py]: why must we use the same reference point? what happens if you don't use the same reference point?

File: trieste/utils/pareto.py (hunk 6 above)
Comment (joelberkeley) [inline trieste/utils/pareto.py:114]: Can you expand on what a reference point is? From the perspective of someone who doesn't know much about multi-objective, or is used to a different name

File: tests/unit/test_pareto.py [hunk 7]
```diff
@@ -111,3 +111,22 @@ def test_pareto_2d_bounds() -> None:
     npt.assert_allclose(
         pareto_2d.front, tf.constant([[0.1419, 0.9340], [0.1576, 0.7922], [0.4854, 0.0357]])
     )
+
+
+def test_hypervolume() -> None:
```
Comment (joelberkeley) [inline tests/unit/test_pareto.py]: the naming convention for tests for methods in trieste is usually `test_<class>_<method>` with `_<what we're testing>` at the end if there's more than one test for that method. This makes it clear what exactly is under test
```suggestion
def test_pareto_hypervolume_indicator() -> None:
```

File: trieste/utils/pareto.py (hunk 1 above)
Comment (satrialoka) [reply trieste/utils/pareto.py]: removed

File: tests/unit/test_pareto.py (hunk 2 above)
Comment (satrialoka) [reply tests/unit/test_pareto.py]: I am using the same value as they use in https://github.com/GPflow/GPflowOpt/blob/3d86bcc000b0367f19e9f03f4458f5641e5dde60/testing/unit/test_pareto.py#L63

File: trieste/utils/pareto.py (hunk 3 above)
Comment (satrialoka) [reply trieste/utils/pareto.py]: The first one, I've updated the docstring.

File: trieste/utils/pareto.py (hunk 4 above)
Comment (satrialoka) [reply trieste/utils/pareto.py]: it's not necessary I think. removed.

File: trieste/utils/pareto.py (hunk 6 above)
Comment (satrialoka) [reply trieste/utils/pareto.py:114]: done

File: trieste/utils/pareto.py (hunk 3 above)
Comment (satrialoka) [reply trieste/utils/pareto.py]: it won't be comparable, since this reference point is used as a point bound to calculate the hypervolume. this is only the case if we want to compare/benchmarking the hypervolume result across runs.

File: trieste/utils/pareto.py (hunk 6 above)
Comment (satrialoka) [reply trieste/utils/pareto.py:114]: this term is specific for hypervolume-based Pareto optimization. mathematically, reference point = max(Pareto front set)+epsilon. epsilon is some constant. visually it is explained in Figure 2 of this article https://link.springer.com/content/pdf/10.1007/s10898-013-0118-2.pdf . 
not sure what to add In the docstring. what if "a bound point to calculate the volume"

File: tests/unit/test_pareto.py (hunk 2 above)
Comment (joelberkeley) [reply tests/unit/test_pareto.py]: do you know how they came across this value?

Comment (joelberkeley) [reply tests/unit/test_pareto.py]: the reason I ask is because it's difficult to tell by looking at this test whether it's right or not

Comment (satrialoka) [reply tests/unit/test_pareto.py]: sorry for late response, I just confirmed it with Ivo. so in the gpflowopt they just try the algorithm (fully with acquisition function and BO steps) on some benchmark function. and see if everything works correctly and use both the pareto and the hypervolume of that pareto (3.3878) given [2, 2] ref point for the test. 
It seems strange way to do it actually.

Comment (satrialoka) [reply tests/unit/test_pareto.py]: one of the option that we can use as Ivo suggest is drawing random points from upper right part of unit circle.  if the reference point is [1,1], the hypervolume indicator should be close to 0.5*0.5 (the area of upper right square) - 1/4 of the circle area.

Comment (joelberkeley) [reply tests/unit/test_pareto.py]: would it not be possible to simply use some hand-chosen pareto points and work out the hypervolume by hand? To take a really simply case, a pareto front of ([-1, -0.5], [-0.5, -1]) with reference [0, 0] would give a hypervolume of 0.75 (do check that - I'm not certain I've got hypervolume yet). Probably worth using more interesting points and reference, but you get my drift

Comment (joelberkeley) [reply tests/unit/test_pareto.py]: we'd also need tests for both an empty pareto front and for a single point. If the expected behaviour for an empty point is undefined, the method should error

File: trieste/utils/pareto.py [hunk 8]
```diff
@@ -105,3 +105,40 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
...
+                (self.bounds.upper_idx, ["N", "D"]),
+                (self.front, ["M", "D"]),
+                (reference, ["D"]),
+            ]
+        )
+
+        min_pfront = tf.reduce_min(self.front, 0, keepdims=True)
+        pseudo_pfront = tf.concat((min_pfront, self.front, reference[None]), 0)
+        D = tf.shape(pseudo_pfront)[1]
+        N = tf.shape(self.bounds.upper_idx)[0]
```
Comment (joelberkeley) [inline trieste/utils/pareto.py]: since you've asserted the shape above, you could write, if you want,
```python
N, D = tf.shape(self.bounds.upper_idx)
```

File: tests/unit/test_pareto.py (hunk 2 above)
Comment (satrialoka) [reply tests/unit/test_pareto.py]: > 
> 
> would it not be possible to simply use some hand-chosen pareto points and work out the hypervolume by hand? To take a really simply case, a pareto front of ([-1, -0.5], [-0.5, -1]) with reference [0, 0] would give a hypervolume of 0.75 (do check that - I'm not certain I've got hypervolume yet). Probably worth using more interesting points and reference, but you get my drift

I don't have any interesting points in mind right now. but your sample indeed pretty easy to visualize and think of. I'll add it. If there are interesting points in the future, we can always add them as additional test points.

Comment (satrialoka) [reply tests/unit/test_pareto.py]: > 
> 
> we'd also need tests for both an empty pareto front and for a single point. If the expected behaviour for an empty point is undefined, the method should error

okay

Comment (joelberkeley) [reply tests/unit/test_pareto.py]: how about I make some tests and send them your way? We should add the tests with the code

Comment (satrialoka) [reply tests/unit/test_pareto.py]: okay

File: trieste/utils/pareto.py (hunk 6 above)
Comment (joelberkeley) [reply trieste/utils/pareto.py:114]: how about "Defines the upper bound of the hypervolume"?

File: trieste/utils/pareto.py [hunk 9]
```diff
@@ -105,3 +105,39 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
...
 
         return BoundedVolumes(lower, upper)
+
+    def hypervolume_indicator(self, reference: TensorType) -> TensorType:
+        """
+        Calculate the hypervolume indicator
+        The hypervolume indicator is the volume of the dominated region.
+
+        :param reference: a reference point to use, with shape [D].
+            Should be equal or bigger than the anti-ideal point of the Pareto set.
```
Comment (joelberkeley) [inline trieste/utils/pareto.py:116]: victor mentioned it's useful to be able to specify a reference point with elements less than those of the anti-ideal point in some dimensions. That this can be used when treating objectives as constraints, or when you want to focus on a particular section of the front.

This algorithm doesn't appear to work in that case. Is that a limitation of the algorithm or implementation?

Comment (joelberkeley) [reply trieste/utils/pareto.py:116]: actually, let's just stick with it as it is, and error if it's not greater than or equal to the anti-ideal point. I think that's just
```
tf.debugging.assert_greater_equal(reference, self.front)
```

Comment (joelberkeley) [reply trieste/utils/pareto.py:116]: we can add the other functionality later

Comment (satrialoka) [reply trieste/utils/pareto.py:116]: okay, maybe better to open an issue for it.

File: trieste/utils/pareto.py (hunk 6 above)
Comment (satrialoka) [reply trieste/utils/pareto.py:114]: sounds good

File: trieste/utils/pareto.py (hunk 9 above)
Comment (joelberkeley) [reply trieste/utils/pareto.py:116]: yeah. I can do that

**Review by joelberkeley — APPROVED** [review]

File: trieste/utils/pareto.py (hunk 9 above)
Comment (satrialoka) [reply trieste/utils/pareto.py:116]: thanks
